### ISSUE TITLE
Evalute 'code smell' usefulness

### ISSUE THREAD

**Issue by Michael1993** [junit-pioneer/junit-pioneer#217]
As can be seen in #216 SonarCloud code smells can often result in "false positives". Should this particular feature be disabled?

Most common false positives are the empty tests we don't actually execute just record with `TestExecutionRecorder`.

**Comment by Bukama** [comment]
At this particular: Definitly as it doesn't fit for our environment.

In General? I don't think so as we then can skip many things. When @aepfli confirms that he has finished setting up sonar then we should evaluate all active rules (not only code smells) and also fix violations of rules we want to keep

**Comment by aepfli** [comment]
Sonar is setup ;)

**Comment by Bukama** [comment]
> Sonar is setup ;)

Okay, I'll try to go through it. Maybe this evening then the others play Stellaris

**Comment by Bukama** [comment]
> > Sonar is setup ;)
> 
> Okay, I'll try to go through it. Maybe this evening then the others play Stellaris

As this said, I'll assign the issue to me as long as I have a look at the checks.

**Comment by Bukama** [comment]
From `src/main`
1. Generic `Exceptions` or `Thrownables` are used, this is correct as you should aways use a particular exception depending on your context, e.g. `IllegalArguementException` for invalid parameters
2. Several local variables don't follow the naming conventions (`^[a-z][a-zA-Z0-9]*$`, e.g. `__` - yeah such names are used! So we have to decide if want these naming convention. As they are the default convetions of Java I would vote to do so
3. `peek()` is used in streams, but offical javadoc says that its for debugging purpose and not to be used in production code
4. `Remove this use of "Test"; it is deprecated` - this is mentioned several times in the vintage extension, as the `@Test` annotation there is deprecated. The rule itself (don't use deprecated classes/methods) is very important in my opinion so we would need to exclude it for this class or change the implementation. This also belongs to the annotation itself (`Do not forget to remove this deprecated code someday`)


From `src/test`
5. `Add at least one assertion to this test case` we got this messages en mass. This is because we mostyl deal with annotations and not with tests itself. I vote for disable this check for pioneer.
6. `Remove this use of "Thread.sleep()"` This is used inside a test for the `TimeoutExtension`. I don't see a way to remove this warning.

I vote for cleaning up all the things from `src/main`. I already cleaned up some things (#221), where I didn't see the need of an discussion. I would do for the points 1 and 2. For 3 I need help as I'm really not into streams (this goes for another violation I don't get fixed, see the draf)

For 4,5,6 I only see the way to disable ( or for 4: ignore / supress) them.

**Comment by aepfli** [comment]
Some issues are `false positives` some definitely not. I agree that 5. makes no sense in our context. and i already mentioned that.

Regarding 6. and `Thread.sleep()` - i think the rule is valid, and sonar is not the last instance of sanity - it is us developers. But in this case we definitely want this, so i would suggest to add a line comment like: `Thread.sleep(): //NOSONAR wanted behaviour for testing` or we could actually use [Awaitility](https://github.com/awaitility/awaitility) instead - but i am not sure if that makes sense for just one test to add this dependency.

Regarding 4.:
We could exclude this folder for the deprecation check, but i am not 100% sure

The other 3 topics are interesting ones, and we should keep them in mind. maybe we should also do a little tutorial on how to add [sonarlint](https://www.sonarlint.org/) to intellij, for live code inspection during development

**Comment by Bukama** [comment]
>>  I agree that 5. makes no sense in our contex

As you are admin: Can you disable the rule in Sonar? 

> Regarding 6. and `Thread.sleep()`

Thanks for the `//NOSONAR` hint. I added this (will push later)

>> Regarding 4.:
Can you add a single exception for the specific class?


> The other 3 topics are interesting ones

I would try to fix them, aside `peek()` I would need help most probably.

**Comment by Bukama** [comment]
To give you an update: I've updated #221 and cleaned the violations at 1 and 2. I'm not sure if all of 3 occurence can be "fixed" and I would need help at this. @aepfli tried to update the sonar rules to get rid of 4. We'll have to see if the exclusions work, after #221 was merged.

**Comment by Bukama** [comment]
After #221 was merged we are down to [three violations](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&resolved=false)

1. Two usages of `peek()`, where I don't know how to replace them
2. One notice to remove the deprecated `@Test` annotation in `vintage`

As for the second: @aepfli Can we create an exclusion here?

**Comment by Michael1993** [comment]
As far as I know the best way to replace `peek()` is with a `forEach()` or a `forEachOrdered()` because these are the methods of the `Stream API` that take a `Consumer` as parameters.

Somewhat like this:
```
List<T> result = collection.stream()
    .peek(anOperation())
    .map(e -> e.getter())
    .collect(toList());
```
into
```
Collection<U> whatever = collection.forEach(anOperation());
List<T> result = whatever.stream()
    .map(e -> e.getter())
    .collect(toList());
```

This is the most straightforward way that I know, however it's ugly. Sometimes a clever `map()` can replace `peek()`, e.g.:
```
peek(value -> System.out.println(value.toString()))
```
can be replaced with
```
map(value -> { System.out.println(value.toString()); return value; })
```
but that might lead to some confusion.

**Comment by Bukama** [comment]
Thank you @Michael1993 I created a PR for this, which should allow us to close this issue after @aepfli  has excluded the deprecated `@Test` from sonar checks

**Comment by aepfli** [comment]
here is the thing with peek. this is merely a warning, as a lot of developers used it, or still using it incorrectly. based on our usecase those are correct. There are three ways of handling that

1. the good old `// NOSONAR`
2. a `@SurpressWarning` for that rule on that method
3. a mark in sonarqube as `False positive` or won't fix.

All of them are legit ways to tell sonar, we know what we are doing. And we are always allowed to do so. There is no need to fix all sonar issues, we just need to evaluate them properly

**Comment by aepfli** [comment]
i marked the two issues as false positives so i think we are good to go to close this issue, arent we?

**Comment by Bukama** [comment]
> i marked the two issues as false positives so i think we are good to go to close this issue, arent we?

I think so.

![sonar](https://user-images.githubusercontent.com/4354359/80787039-aa29d180-8b85-11ea-9228-f668deb61ffd.png)
